gu-client/src/async.rs

```diff
@@ -0,0 +1,297 @@
+                url: format!("http://{}/", addr.into()),
+            }),
+        }
+    }
+    /// creates a new hub session
+    /// assumption: url is /sessions, hub session is in the response body
+    pub fn new_session(
+        &self,
+        session_info_builder: SessionInfoBuilder,
+    ) -> impl Future<Item = HubSession, Error = Error> {
+        let sessions_url = format!("{}sessions", self.driver_inner.url);
+        let session_info = match session_info_builder.build() {
+            Ok(r) => r,
+            _ => return future::Either::A(future::err(Error::InvalidHubSessionParameters)),
+        };
+        let request = match client::ClientRequest::post(sessions_url).json(session_info) {
+            Ok(r) => r,
+            _ => return future::Either::A(future::err(Error::CannotCreateRequest)),
+        };
+        let driver_for_session = self.clone();
+        future::Either::B(
+            request
+                .send()
+                .map_err(|_| Error::CannotSendRequest)
+                .and_then(|response| {
+                    if response.status().as_u16() != 201 {
+                        return future::Either::A(future::err(Error::CannotCreateSession));
+                    }
+                    future::Either::B(response.body().map_err(|_| Error::CannotGetResponseBody))
+                }).and_then(|body| {
+                    future::ok(HubSession {
+                        driver: driver_for_session,
+                        session_id: match str::from_utf8(&body.to_vec()) {
+                            Ok(str) => str.to_string(),
+                            _ => return future::err(Error::CannotGetResponseBody),
+                        },
+                    })
+                }),
+        )
+    }
+    pub fn auth_app<T: Into<String>, U: Into<String>>(&self, _app_name: T, _token: Option<U>) {}
+    /// returns all peers connected to the hub
+    /// assumption: url is /peer
+    pub fn list_peers(&self) -> impl Future<Item = impl Iterator<Item = PeerInfo>, Error = Error> {
+        let url = format!("{}peer", self.driver_inner.url);
+        return match client::ClientRequest::get(url).finish() {
+            Ok(r) => future::Either::A(
+                r.send()
+                    .map_err(|_| Error::CannotSendRequest)
+                    .and_then(|response| match response.status().as_u16() {
+                        200 => future::Either::A(
+                            response.json().map_err(|_| Error::InvalidJSONResponse),
+                        ),
+                        _ => future::Either::B(future::err(Error::InternalError)),
+                    }).and_then(|answer_json: Vec<PeerInfo>| future::ok(answer_json.into_iter())),
+            ),
+            _ => future::Either::B(future::err(Error::CannotCreateRequest)),
+        };
+    }
+}
+
+/// Hub session.
+#[derive(Clone, Debug)]
+pub struct HubSession {
+    driver: Driver,
+    pub session_id: String,
```

Comment:
`session_id` field probably should be private. Hub session has a permanent id so it should be set on creation and then immutable.

Reply:
Fixed.